test/y2firstboot/clients/user_test.rb

```diff
@@ -0,0 +1,139 @@
+    let(:system_config_copy) { Y2Users::Config.new }
+    let(:config_manager)     { Y2Users::ConfigManager.instance }
+
+    let(:writer) { instance_double(Y2Users::Linux::Writer, write: []) }
+
+    before do
+      user.password = password
+      system_config_copy.attach([user])
+
+      allow(Yast::InstUserFirstDialog).to receive(:new).and_return(dialog)
+
+      allow(Y2Users::Linux::Writer).to receive(:new).and_return(writer)
+
+      allow(subject).to receive(:user).and_return(user)
+      allow(user).to receive(:attached?).and_return(attached)
+
+      allow(system_config).to receive(:copy).and_return(system_config_copy)
+      allow(config_manager).to receive(:system).and_return(system_config)
+    end
+
+    context "when user has an encrypted password" do
+      let(:password) { Y2Users::Password.create_encrypted("s3cr3t") }
+
+      it "resets the user password" do
+        expect(user.password.value).to be_encrypted
+
+        subject.run
+
+        expect(user.password.value).to_not be_encrypted
+        expect(user.password_content).to be_empty
+      end
+    end
+
+    context "when user has a plain password" do
+      let(:password) { Y2Users::Password.create_plain("s3cr3t") }
+
+      it "does not reset the user password" do
+        expect(user.password_content).to eq("s3cr3t")
+
+        subject.run
+
+        expect(user.password_content).to eq("s3cr3t")
+      end
+    end
+
+    it "executes the inst_user_first dialog" do
+      expect(Yast::InstUserFirstDialog).to receive(:new).with(system_config_copy, user: user)
+      expect(dialog).to receive(:run)
+
+      subject.run
+    end
+
+    it "returns the dialog result" do
+      expect(subject.run).to eq(result)
+    end
+
+    context "when dialog result is :next" do
+      it "writes the users configuration" do
+        expect(Y2Users::Linux::Writer).to receive(:new).with(system_config_copy, system_config)
+        expect(writer).to receive(:write)
+
+        subject.run
+      end
+
+      context "and user still attached to the config" do
+        let(:attached) { true }
+
+        it "saves the username for future reference" do
+          expect(described_class).to receive(:username=).with(username)
+
+          subject.run
+        end
+      end
+
+      context "but user is not attached to the config" do
+        let(:attached) { false }
+
+        it "deletes username reference" do
+          expect(described_class).to receive(:username=).with(nil)
+
+          subject.run
+        end
+      end
+    end
+
+    context "when dialog result is not :next" do
+      let(:result) { :back }
+
+      it "does not modify stored username" do
+        expect(described_class).to_not receive(:username=)
+
+        subject.run
+      end
+
+      it "does not write the users configuration" do
+        expect(Y2Users::Linux::Writer).to_not receive(:new)
+        expect(writer).to_not receive(:write)
+
+        subject.run
+      end
```

Comment:
Should it delete the username reference? If so, I would add a test.

Reply:
No, it should not. Think that maybe you're going :back and then you have to going :next again. So, it would be nice to keep the username reference until user is detached from configuration.

Reply:
Umm, I was thinking when the dialog result is `:skip`.

Reply:
> Umm, I was thinking when the dialog result is `:skip`.

:thinking: right!